app/roles.go

```diff
@@ -142,10 +142,30 @@ func (c *RolesChanges) Adjust(leader uint64) (client.NodeRole, []client.NodeInfo
 		return -1, nil
 	}
 
-	onlineVoters := c.list(client.Voter, true)
-	onlineStandbys := c.list(client.StandBy, true)
-	offlineVoters := c.list(client.Voter, false)
-	offlineStandbys := c.list(client.StandBy, false)
+	onlineVoters := c.list(client.Voter, true, nil)
+	onlineStandbys := c.list(client.StandBy, true, nil)
+	offlineVoters := c.list(client.Voter, false, nil)
+	offlineStandbys := c.list(client.StandBy, false, nil)
+
+	domainsWithVoters := c.failureDomains(onlineVoters)
+	allDomains := c.allFailureDomains()
+
+	// If we do not have voters on all failure domains and we have a domain with more than one voters
+	// we may need to send voters to domains without voters.
+	if len(domainsWithVoters) < len(allDomains) && len(domainsWithVoters) < len(onlineVoters) {
+		// Find the domains we need to populate with voters
+		domainsWithoutVoters := c.domainsSubtract(allDomains, domainsWithVoters)
+		// Find StandBys in the domains we need to populate
+		candidates := c.list(client.StandBy, true, domainsWithoutVoters)
+		// If we do not have StandBys to promote see if we have Spares
+		if len(candidates) <= 0 {
+			candidates = c.list(client.Spare, true, domainsWithoutVoters)
+		}
+		if len(candidates) > 0 {
+			c.sortCandidates(candidates, domainsWithoutVoters)
+			return client.Voter, candidates
+		}
```

Comment:
The intention here is to favor the promotion of a StandBy node over a Spare. Promoting a StandBy node should be faster than a Spare node since we do not need to transfer the DB, right?

It makes me wonder if the promotion to a voter in https://github.com/canonical/go-dqlite/blob/master/app/roles.go#L160 should also favor StandBys. It does not feel right to have StandBys but when we want to promote a node we consider Spare and StandBys under equal terms.

Reply:
In https://github.com/canonical/go-dqlite/blob/master/app/roles.go#L158-L169 we create a candidates list with the StandBys first and the Spares last, then we sort the list based on the domain and the weight of each candidate. If I understand this correctly the sorting of the list does not respect our intention to prioritize StandBys over Spares and as a result we could end up promoting a Spare to a Voter even if we already have a StandBy.

Reply:
Maybe on the long run it makes more sense to prioritize first based on the domain then on the weight and then on the node type.

Reply:
The latest code follows your suggestion of considering both Sapre and StandBy nodes for promotion. Let's address my concerns on the promotion criteria on a separate PR.